### PR TITLE
Several fixes to pkg-config dependencies and the pkg-config module (try 3)

### PR DESCRIPTION
**Pull request by nirbheek**
`dependency(..., static: true)` had little effect when using pkg-config dependencies because we still passed `-lfoo -lbaz` to the compiler which with gcc meant that you would use shared libraries, and with MSVC meant that we would convert that to `foo.lib baz.lib` which meant that we would use import libraries (when linking to projects built with Meson) or possibly static libraries (those that were not built with Meson, because we name them `libfoo.a`).

With these changes, when `static: true` is set, we parse the pkg-config output, and if it contains `-L` fields, we use `cc.find_library`'s manual library searching feature to find the actual location of the library and use it directly.

These changes also improve `cc.find_library`'s manual searching to only search for platform-specific libraries. Earlier, it would find `libfoo.dylib` as a valid file on non-macOS platforms.

You can now also specify what type of library you want to find. For instance, you can specify that you only want static libraries or only shared libraries or prefer static over shared, etc. This functionality is currently not exposed to build files because it only works when `extra_dirs` is specified. However, I can see it being useful in the future because this is an oft-asked-for feature among autotools and libtool users.

We now also use `shlex.split` to parse `Libs:` and `Cflags:` in pkg-config files so that we properly parse them when `libdir`, `prefix`, etc contain spaces. I also found that our pkg-config generator does not properly escape paths with spaces (`prefix`, `libdir`, etc), which is [required by all implementations](https://github.com/pkgconf/pkgconf/issues/153), so I also fixed that.

Overall, pkg-config usage on Windows should be much better with these commits.

### CODE REVIEW

**Comment by nirbheek** [comment]
@jon-turney the problem here seems to be that the shared-lib unit test never ran with the cygwin toolchain; on cygwin it was using msvc. Now that we do use it, it fails. I'm not sure why it's not working, can you take a look at it?

**Comment by jon-turney** [comment]
@nirbheek https://github.com/jon-turney/meson/commit/4d7e822af4c8687f772d4fc1880070455cbfcab2 seems enough to fix this.

> @jon-turney the problem here seems to be that the shared-lib unit test never ran with the cygwin toolchain; on cygwin it was using msvc.

I'm pretty sure it wasn't using MSVC.  I think it was using the MSVC naming conventions.

File: .appveyor.yml [hunk 1]
```diff
@@ -89,6 +89,8 @@ install:
   # but they have a broken certificate (on Windows) and an HSTS policy that
   # forces usage of HTTPS, so we mirror our own copy.
   - ps: If($Env:compiler -eq 'msys2-mingw') {(new-object Net.WebClient).DownloadFile('https://nirbheek.in/files/meson/get-pip.py', 'C:\projects\meson\get-pip.py')}
+  # pkg-config is needed for some msvc and cygwin tests
```
Comment (jon-turney) [inline .appveyor.yml]: It's misleading to mention cygwin in this comment.  We're not installing or using this pkg-config for cygwin, nor should we be.

**Comment by nirbheek** [comment]
Thanks! I'll add that to this PR. Not sure how I missed it. :)

File: .appveyor.yml (hunk 1 above)
Comment (nirbheek) [reply .appveyor.yml]: Hmm, I think I had to download this for cygwin too. Will try without it.

Comment (jon-turney) [reply .appveyor.yml]: I hope the condition `If($Env:compiler -ne 'msys2-mingw')` means this line isn't doing anything on cygwin, currently.

Comment (nirbheek) [reply .appveyor.yml]: I've changed it; it only installs for MSVC now.

**Comment by jpakkane** [comment]
The prebuilt test failure seems very relevant.

**Comment by nirbheek** [comment]
It seems the cygwin patch fixed cygwin and broke everything else. Looking into it.

**Comment by jon-turney** [comment]
Oops.  Should be `elif` not `if`

File: run_unittests.py [hunk 2]
```diff
@@ -1348,6 +1348,9 @@ def detect_prebuild_env(self):
         if mesonbuild.mesonlib.is_windows():
             object_suffix = 'obj'
             shared_suffix = 'dll'
+        if mesonbuild.mesonlib.is_cygwin():
```
Comment (jon-turney) [inline run_unittests.py]: elif :blush:

**Comment by nirbheek** [comment]
> Oops. Should be elif not if

That shouldn't make a difference right, unless `is_cygwin()` is returning `True` when we're on Windows but not on Cygwin?

**Comment by jon-turney** [comment]
There's an else: clause at the end which will now get run on windows...

**Comment by nirbheek** [comment]
Oh right. I knew I shouldn't be reading patches at 3am.

**Comment by nirbheek** [comment]
@jon-turney cygwin is broken again, this is so funny :D

**Comment by jon-turney** [comment]
@nirbheek this looks like an unrelated sporadic failure.

Probably worth digging into, but not much information to go on in the log...

**Comment by nirbheek** [comment]
Yep, it passed this time. Maybe it's worth running the test suite in a loop on your cygwin system to catch any weird racy issues? I also wonder how much of the sporadic test failures are because of antiviruses.

File: run_unittests.py [hunk 3]
```diff
@@ -1436,6 +1453,68 @@ def test_prebuilt_shared_lib(self):
                     if os.path.splitext(fname)[1] not in ['.c', '.h']:
                         os.unlink(fname)
 
+    def test_pkgconfig_static(self):
+        '''
+        Test that the we only use static libraries when `static: true` is
+        passed to dependency() with pkg-config. Can't be an ordinary test
+        because we need to build libs and try to find them from meson.build
+        '''
```
Comment (jpakkane) [inline run_unittests.py]: Do we really need these justification comments in every unit test? It is fairly obvious (or should be during code review) why the tests are run.

This is not so much for this MR but for future ones.

Comment (nirbheek) [reply run_unittests.py]: It always makes sense what a unit test does to core contributors, but it makes no sense to new contributors. They're not really *needed*, but I find that they're useful in making the codebase more approachable.

**Comment by tp-m** [comment]
I often find myself looking at unit tests other people wrote (or sometimes even some that I wrote years ago myself), and wondering what in the world these are trying to test, and it's not always easy to figure out even for a codebase I have been working on for more than a decade, so I think these explanatory comments are great. They will also help later to figure out failures and behavioural changes, as it's easier to see whether a test doesn't actually test exactly what it was meant to test (e.g. because it was making assumptions that may or may not be true any more now) or whether it is behaving as intended and something actually broke.

**Comment by mojca** [comment]
These changes break the cross-compiler (#2659). I didn't bisect any further yet, but the problem indeed appears somewhere between 8e04ab7cf2d16830e62e3afbfe9e7f20a5b27d46 (works) and 9da031c04c9905949c896a8fdc5b608acd6dde47 (doesn't work).

**Pull request merged**